Content.Client/GameObjects/Components/CloningPod/CloningPodWindow.cs

```diff
@@ -0,0 +1,442 @@
+
+            // Create buttons at the start of the list that are now visible (scrolling up).
+            for (var i = Math.Min(prevStart - 1, endIndex); i >= startIndex; i--)
+            {
+                InsertEntityButton(array[i], true, i);
+            }
+
+            // Create buttons at the end of the list that are now visible (scrolling down).
+            for (var i = Math.Max(prevEnd + 1, startIndex); i <= endIndex; i++)
+            {
+                InsertEntityButton(array[i], false, i);
+            }
+        }
+
+        // Create a spawn button and insert it into the start or end of the list.
+        private void InsertEntityButton(KeyValuePair<int, string> scan, bool insertFirst, int index)
+        {
+            var button = new CloningScanButton
+            {
+                Scan = scan.Value,
+                Id = scan.Key,
+                Index = index // We track this index purely for debugging.
+            };
+            button.ActualButton.OnToggled += OnItemButtonToggled;
+            var entityLabelText = scan.Value;
+
+            button.EntityLabel.Text = entityLabelText;
+
+            if (scan.Key == SelectedScan)
+            {
+                _selectedButton = button;
+                _selectedButton.ActualButton.Pressed = true;
+            }
+
+            //TODO: replace with body's face
+            /*var tex = IconComponent.GetScanIcon(scan, resourceCache);
+            var rect = button.EntityTextureRect;
+            if (tex != null)
+            {
+                rect.Texture = tex.Default;
+            }
+            else
+            {
+                rect.Dispose();
+            }
+
+            rect.Dispose();
+            */
+
+            _scanList.AddChild(button);
+            if (insertFirst)
+            {
+                button.SetPositionInParent(0);
+            }
+        }
+
+        private static bool _doesScanMatchSearch(string scan, string searchStr)
+        {
+            return scan.ToLowerInvariant().Contains(searchStr);
+        }
+
+        private void OnItemButtonToggled(BaseButton.ButtonToggledEventArgs args)
+        {
+            var item = (CloningScanButton) args.Button.Parent!;
+            if (_selectedButton == item)
+            {
+                _selectedButton = null;
+                SelectedScan = null;
+                return;
+            }
+            else if (_selectedButton != null)
+            {
+                _selectedButton.ActualButton.Pressed = false;
+            }
+
+            _selectedButton = null;
+            SelectedScan = null;
+
+            _selectedButton = item;
+            SelectedScan = item.Id;
+        }
+
+        protected override void FrameUpdate(FrameEventArgs args)
+        {
+            base.FrameUpdate(args);
+            UpdateVisibleScans();
+        }
+
+        private class ScanListContainer : Container
+        {
+            // Quick and dirty container to do virtualization of the list.
```

Comment:
I'd prefer a proper virtual list control, tbh.
Adhoc implementations like this aren't great.

You don't have to do this in this PR (obviously, atomic), but perhaps you could generalise your implementation and commit it?

Reply:
Definitely, although I actually ripped off the EntitySpawnWindow.cs for most of it. CargoConsoleMenu could probably also benefit from this.

Reply:
Ah!! see! that proves my point, adhoc little implementations end up being copied and sprinkled around as necessary :P